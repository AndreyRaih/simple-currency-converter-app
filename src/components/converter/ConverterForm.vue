<template>
  <converter-controlls
    v-model:amount="converterFormData.amount"
    v-model:current="converterFormData.from"
    v-model:target="converterFormData.to"
    :symbols="symbols"
    :loading="loading"
    @switch-currencies="handleSwitchCurrencies"
    @convert-currencies="() => $emit('convert', converterFormData)"
  />
  <n-collapse-transition :show="showResultView">
    <converter-result :data="converterResult" />
  </n-collapse-transition>
</template>

<script lang="ts">
import { defineComponent, computed, PropType, ref, watch, onBeforeMount, onBeforeUnmount } from 'vue'
import ConverterControlls from './views/Controlls.vue'
import ConverterResult from './views/Result.vue'
import { NCollapseTransition } from 'naive-ui'
import type { SelectOption } from 'naive-ui'
import type { ConversionResult, ConversionQuery } from '@/typings.d'

export default defineComponent({
  name: 'ConverterView',
  components: {
    ConverterControlls,
    ConverterResult,
    NCollapseTransition
  },
  props: {
    converterData: {
      required: true,
      type: Object as PropType<ConversionQuery>,
      default: () => ({
        from: '',
        to: '',
        amount: 0
      })
    },
    converterResult: {
      required: true,
      type: Object as PropType<ConversionResult>,
      default: null
    },
    symbols: {
      required: true,
      type: Array as PropType<SelectOption[]>
    },
    loading: {
      type: Boolean,
      default: false
    }
  },
  emits: [
    'convert',
    'clear-result'
  ],
  setup (props, { emit }) {
    const converterFormData = ref<ConversionQuery>({
      from: '',
      to: '',
      amount: 0
    })

    const showResultView = computed(() => Boolean(props.converterResult && Object.keys(props.converterResult).length))

    watch(() => converterFormData.value, () => emit('clear-result'), { deep: true })

    onBeforeMount(() => {
      /**
       * NOTE: Here we break the reactivity bindings. It has been done for 2 reasons:
       * 1. Separate inner objects from outer, and avoid mutating source state
       * 2. Stick to "Data - Down, Actions - Up" approach
       * 
       * Actually, if i got it right, for now there is no better way to handle similar cases
       */
      converterFormData.value = { ...props.converterData }
    })

    onBeforeUnmount(() => emit('clear-result'))

    const handleSwitchCurrencies = () => {
      const prevTo = converterFormData.value.to
      const prevFrom = converterFormData.value.from
      converterFormData.value.from = prevTo
      converterFormData.value.to = prevFrom
    }

    return {
      converterFormData,
      handleSwitchCurrencies,
      showResultView
    }
  }
})
</script>
