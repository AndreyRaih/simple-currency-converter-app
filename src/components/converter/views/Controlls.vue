<template>
  <section :class="$style.Container">
    <n-form-item :class="$style.FormItem" :label-style="currencyFieldLabelStyle" label="Amount">
      <n-input-number
        :class="$style.AmountInput"
        :value="amount"
        placeholder="Enter amount"
        @update:value="(value) => $emit('update:amount', value as number)"
      />
    </n-form-item>

    <currency-select
      :class="$style.FormItem"
      label="From"
      placeholder="Select From"
      :options="symbols"
      :value="current"
      @update:value="(value) => $emit('update:current', value)"
    />

    <n-button :class="$style.ReplaceButton" circle @click="() => $emit('switch-currencies')">
      <template #icon>
        <n-icon>
          <replace-icon />
        </n-icon>
      </template>
    </n-button>

    <currency-select
      :class="$style.FormItem"
      label="To"
      placeholder="Select To"
      :options="symbols"
      :value="target"
      @update:value="(value) => $emit('update:target', value)"
    />

    <n-button
      :class="$style.ConvertButton"
      :loading="loading"
      type="primary"
      @click="() => $emit('convert-currencies')"
    >Convert</n-button>
  </section>
</template>

<style lang="less" module>
.Container {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
}

.FormItem {
  flex: 1;
  width: 100%;
  min-width: 160px;
}

.ReplaceButton {
  margin: 26px 14px 0 14px;
}

.ConvertButton {
  margin: 26px 0 0 14px;
}

.AmountInput {
  width: 90%;
}

@media screen and (max-width: 768px) {
  .Container {
    flex-direction: column;
    margin-bottom: 48px;
  }

  .ReplaceButton {
    margin: 0;
    margin-bottom: 24px;
  }

  .ConvertButton {
    margin: 0;
    margin-top: 24px;
  }

  .AmountInput {
    width: 100%;
  }
}
</style>

<script lang="ts">
import { defineComponent, PropType } from 'vue'
import { NFormItem, NInputNumber, NButton, NIcon } from 'naive-ui'
import { ArrowsHorizontal as ReplaceIcon } from '@vicons/carbon'
import type { SelectOption } from 'naive-ui'
import CurrencySelect from '@/components/shared/CurrencySelect.vue'
import { currencyFieldLabelStyle } from '@/components/shared/constants'

export default defineComponent({
  name: 'ConverterAmountForm',
  components: {
    NFormItem,
    NInputNumber,
    CurrencySelect,
    NButton,
    NIcon,
    ReplaceIcon
  },
  props: {
    amount: {
      type: Number || String,
      required: true,
      default: null
    },
    current: {
      type: String,
      required: true,
      default: null
    },
    target: {
      type: String,
      required: true,
      default: null
    },
    symbols: {
      type: Array as PropType<SelectOption[]>,
      required: true,
      default: () => ([])
    },
    loading: {
      type: Boolean,
      default: false
    }
  },
  emits: [
    'update:amount',
    'update:current',
    'update:target',
    'switch-currencies',
    'convert-currencies'
  ],
  setup () {
    return {
      currencyFieldLabelStyle
    }
  }
})
</script>
