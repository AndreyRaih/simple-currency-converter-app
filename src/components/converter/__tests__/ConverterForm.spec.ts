import { shallowMount } from '@vue/test-utils'
import ConverterForm from '../ConverterForm.vue'
import Controlls from '../views/Controlls.vue'

const defaultOptions = {
  props: {
    converterData: {
      from: 'USD',
      to: 'EUR',
      amount: 0
    },
    converterResult: {},
    symbols: [],
    loading: false
  }
}

describe('ConverterForm.vue', () => {
  describe('Rendering', () => {
    it('should renders correctly', () => {
      const wrapper = shallowMount(ConverterForm, defaultOptions)
      expect(wrapper.element).toMatchSnapshot()
    })

    it('should renders result section, if converterResult is not falsy or empty', async () => {
      const wrapper = shallowMount(ConverterForm, defaultOptions)
      await wrapper.setProps({ converterResult: { baseAmount: 0, targetAmount: 0 }})
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('Functional', () => {
    it('props.converterData should be immutable', () => {
      const wrapper = shallowMount(ConverterForm, defaultOptions)
      wrapper.vm.converterFormData.amount++

      expect(wrapper.vm.converterFormData.amount === wrapper.vm.converterData.amount).toBeFalsy()
    })

    it('custom event @switch-currencies should be handled correct', () => {
      const wrapper = shallowMount(ConverterForm, defaultOptions)

      const handleSwitchCurrenciesTriggerElement = wrapper.findComponent(Controlls)
      handleSwitchCurrenciesTriggerElement.vm.$emit('switch-currencies')

      expect(wrapper.vm.converterFormData.from).toBe(defaultOptions.props.converterData.to)
      expect(wrapper.vm.converterFormData.to).toBe(defaultOptions.props.converterData.from)
    })

    it('custom event @convert-currencies should be handled correct', () => {
      const wrapper = shallowMount(ConverterForm, defaultOptions)

      const handleSwitchCurrenciesTriggerElement = wrapper.findComponent(Controlls)
      handleSwitchCurrenciesTriggerElement.vm.$emit('convert-currencies')

      const convertEvent = wrapper.emitted('convert');
      expect(convertEvent).toHaveLength(1)
      expect((convertEvent as unknown[])[0]).toEqual([defaultOptions.props.converterData])
    })
  })
})
