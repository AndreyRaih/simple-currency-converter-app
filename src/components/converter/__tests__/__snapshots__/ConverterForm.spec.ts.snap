// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConverterForm.vue Rendering should renders correctly 1`] = `
<div
  data-v-app=""
>
  
  <converter-controlls-stub
    amount="0"
    current="USD"
    loading="false"
    symbols=""
    target="EUR"
  />
  <n-collapse-transition-stub
    appear="false"
    show="false"
  />
  
</div>
`;

exports[`ConverterForm.vue Rendering should renders result section, if converterResult is not falsy or empty 1`] = `
<div
  data-v-app=""
>
  
  <converter-controlls-stub
    amount="0"
    current="USD"
    loading="false"
    symbols=""
    target="EUR"
  />
  <n-collapse-transition-stub
    appear="false"
    show="true"
  />
  
</div>
`;
